src/type/definition.js

```diff
@@ -1587,3 +1597,61 @@ export function isRequiredInputField(
 }
 
 export type GraphQLInputFieldMap = ObjMap<GraphQLInputField>;
+
+/**
+ * Historically GraphQL.js allowed default values to be provided as
+ * assumed-coerced "internal" values, however default values should be provided
+ * as "external" pre-coerced values. `uncoerceDefaultValue()` will convert such
+ * "internal" values to "external" values to avoid breaking existing clients.
+ *
+ * This performs the opposite of `coerceInputValue()`. Given an "internal"
+ * coerced value, reverse the process to provide an "external" uncoerced value.
+ *
+ * This function should not throw Errors on incorrectly shaped input. Instead
+ * it will simply pass through such values directly.
+ *
+ */
+function uncoerceDefaultValue(value: mixed, type: GraphQLInputType): mixed {
+  // Explicitly return the value null.
+  if (value === null) {
+    return null;
+  }
+
+  // Unwrap type
+  const namedType = getNamedType(type);
+
+  if (isIterableObject(value)) {
+    return Array.from(value, (itemValue) =>
+      uncoerceDefaultValue(itemValue, namedType),
+    );
+  }
+
+  if (isInputObjectType(namedType)) {
+    if (!isObjectLike(value)) {
+      return value;
+    }
+
+    const fieldDefs = namedType.getFields();
+    return mapValue(value, (fieldValue, fieldName) =>
+      fieldName in fieldDefs
+        ? uncoerceDefaultValue(fieldValue, fieldDefs[fieldName].type)
+        : fieldValue,
+    );
+  }
+
+  if (isLeafType(namedType)) {
+    try {
+      // For leaf types (Scalars, Enums), serialize is the oppose of coercion
+      // (parseValue) and will produce an "external" value.
+      return namedType.serialize(value);
```

Comment:
as outlined in a previous comment I think this is wrong: we are misusing `serialize` here or produce a representation for an input value. We should introduce a specific function `inputValueToAst` for the Coercion and leave it to the scalar itself to determine a printable representation of its input values.

Reply:
We talked about this, but writing it down here for posterity:

I agree this is a misuse. This "uncoerce" step should ideally not exist, and only does so to avoid introducing a breaking change to existing users. 

The previous code path used `astFromValue(defaultValue)` - `astFromValue` misuses `serialize` in the same way. This carries over the misuse to preserve existing behavior but isolates it to this specific case.

The new code path is `valueToLiteral(uncoerceDefaultValue(defaultValue))`

Reply:
I am not sure this will always preserve backwards compatibility.

Some scalars throw on serialize as a way of preventing them from being erroneously used as output values. While the example I know of should never have a default value, other examples might. See https://github.com/jaydenseric/graphql-upload/issues/194

I would suggest that as part of the definitive solution that you appear on the cusp of delivering:
(A) the spec and the implementation should provide a mechanism of labeling leaf values as only valid within either output or input values

Separately, I would suggest:
(B) the implementation should introduce programmatic internalDefaultValue/externalDefaultValue options for arguments where only one may be defined, with a deprecation notice for the use of the previously ambiguous defaultValue. This would preserve the current behavior as best as possible, but also provide a clear upgrade path.

Reply:
> I am not sure this will always preserve backwards compatibility.
>
> Some scalars throw on serialize as a way of preventing them from being erroneously used as output values.

Since the previous codepath used `valueToAST` which also called `serialize()` on these, I believe such a scalar would already exhibit this behavior, so no change. This just moves the call to `serialize()`.

> (A) the spec and the implementation should provide a mechanism of labeling leaf values as only valid within either output or input values

This is an interesting idea! It's out of scope for this RFC, but I see the use case for giving GraphQL validation this information, rather than waiting to fail at execution time.

> (B) the implementation should introduce programmatic internalDefaultValue/externalDefaultValue options for arguments where only one may be defined, with a deprecation notice for the use of the previously ambiguous defaultValue. This would preserve the current behavior as best as possible, but also provide a clear upgrade path.

I agree with this! I'd like for this "uncoerce" method to have a reasonably short shelf life and to be able to be removed in a future major version, but I'm looking for a solution that allows the end state to look correct. IMHO having to provide `externalDefaultValue` for the foreseeable future would be too confusing.

I'm not opposed to introducing a more modest breaking change, but would look to @IvanGoncharov for guidance on when and how to introduce that. Some ideas:

1. Require `defaultValue` to validate as an external input value, but provide a new `deprecatedCoercedDefaultValue` field which uses this new method for preserving behavior. 

  Any problematic internal defaultValue would need to manually edit their schema to either fix the value or use this new field. The schema validation error message could reference this field so fixing an existing schema is self-documenting.
2. Provide a "flag" like "allowDeprecatedCoercedDefaultValues" when constructing a schema which applies the behavior currently in this PR. That would be a much easier change to make (one place, instead of per-call-site) but doesn't help highlight the long-term necessary changes and unnecessarily would run "uncoerce" on already uncoerced values.

I'll note that *most* defaultValue tend to be simple built-in scalars, so it's very likely that the vast majority will upgrade to a version including this change and not need to do anything at all. defaultValue for Enum or some Custom Scalars will need to be addressed.

Reply:
I think option 1 is best and imho that it is very important to fix the reference implementation to reflect the spec properly.

Thanks for your additional comments above, much appreciated.